src/Features/Core/Portable/Completion/CommonCompletionService.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal abstract partial class CommonCompletionService : CompletionServiceWithProviders
+    {
+        protected CommonCompletionService(Workspace workspace)
+            : base(workspace)
+        {
+        }
+
+        // Magic trigger for VB snippets
+        internal static readonly CompletionTriggerKind SnippetsTriggerReason = (CompletionTriggerKind)(-1);
+
+        protected override ImmutableArray<CompletionProvider> GetProviders(ImmutableHashSet<string> roles, CompletionTrigger trigger)
+        {
+            var providers = base.GetProviders(roles, trigger);
+
+            if (this.Language == LanguageNames.VisualBasic)
```

Comment:
Common layer should never access LanguageNames.CSharp or VisualBasic.  Can we have some sort of ILanguageService that exposes this flexibility.